### PR TITLE
Adding cloning, cloning program, cloning machine, and adjusted stack …

### PR DESCRIPTION
**Pull request by biscuitWizard**
…to new life flow.

### CODE REVIEW

File: mods/persistence/datums/mind/mind.dm [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+		var/clone_pod_tag = input(usr, "Choose a cloning pod to awaken at:", "Select Cloning Pod") as null|anything in valid_clone_pods
+		if(clone_pod_tag)
+			var/datum/extension/network_device/cloning_pod/CP = valid_clone_pods[clone_pod_tag]
+			if(!CP.is_valid_respawn(unique_id))
+				to_chat(usr, SPAN_WARNING("That cloning pod has become unavailable. Please choose a new cloning pod."))
+				return show_valid_respawns(ckey)
+			CP.create_character(src, ckey)
+			philotic_damage += rand(3, 15)
+			if(philotic_damage > 75)
+				to_chat(usr, SPAN_WARNING("Colors seem to lack the vibrance they used to have. It would be easy to just go to sleep and never wake up."))
```
Comment (NataKilar) [inline mods/persistence/datums/mind/mind.dm]: You could literally add a client color here to reduce the vibrance of colors. A neat flavor thing

Comment (biscuitWizard) [reply mods/persistence/datums/mind/mind.dm]: Oh shit, what would that look like?

File: mods/persistence/datums/mind/mind.dm [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	. = ..()
+	unique_id = "[sequential_id("/datum/mind")]"
+
+/proc/get_valid_clone_pods(var/mind_id)
+	var/list/valid_clone_pods = list()
+	for(var/network_id in GLOB.computer_networks)
+		var/datum/computer_network/network = GLOB.computer_networks[network_id]
+		for(var/datum/extension/network_device/cloning_pod/CP in network.devices)
+			if(CP.is_valid_respawn(mind_id))
+				valid_clone_pods["Backup (DATEGOHERE) - (AREA NAME HERE)"] = CP
```
Comment (Andrew-Fall) [inline mods/persistence/datums/mind/mind.dm]: Looks like placeholder text

File: mods/persistence/game/machinery/resleever.dm [hunk 3]
```diff
@@ -182,7 +182,7 @@ obj/machinery/resleever/Process()
 		eject_occupant()
 
 /obj/machinery/resleever/proc/sleeve()
-	if(lace && !lace.prompting && occupant) // Not only check for the lace and occupant, but also if the lace isn't already prompting the dead user.
+	if(lace && occupant) // Not only check for the lace and occupant, but also if the lace isn't already prompting the dead user.
```
Comment (Andrew-Fall) [inline mods/persistence/game/machinery/resleever.dm]: Comment needs update

File: mods/persistence/modules/cloning/backup_file.dm [hunk 4]
```diff
@@ -0,0 +1,31 @@
+#define BASE_CLONE_BACKUP_SIZE 10
+
+/datum/computer_file/data/cloning
+	filetype = "CLN"
+	read_only = TRUE
+	var/mob_age			// Mob's age in years.
```
Comment (Andrew-Fall) [inline mods/persistence/modules/cloning/backup_file.dm:6]: Might be better to save the mobs birthdate? Or maybe save that for a seperate PR that makes players able to select a birthday and such.

File: mods/persistence/modules/mob/death.dm [hunk 5]
```diff
@@ -0,0 +1,55 @@
+/mob/living/carbon/death()
+	if(stat & DEAD)
+		return FALSE
+
+	to_world_log("death called?")
```
Comment (Andrew-Fall) [inline mods/persistence/modules/mob/death.dm]: Debug message

File: mods/persistence/modules/mob/death.dm [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+	if(ckey && client)
+		to_world_log("player died")
```
Comment (Andrew-Fall) [inline mods/persistence/modules/mob/death.dm]: Debug message

File: mods/persistence/modules/mob/death.dm [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+		var/obj/item/organ/internal/stack/stack = switchToStack(mind, ckey)
+		if(stack)
+			to_world_log("stack found")
```
Comment (Andrew-Fall) [inline mods/persistence/modules/mob/death.dm]: Debug message

File: mods/persistence/modules/mob/death.dm [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+			// They did have a stack.
+			to_chat(stack.stackmob, SPAN_NOTICE("Darkness envelopes you. Your character has died and you are now in limbo. Resleeve to continue playing as your character, or wait until a kind soul clones you from your cortical stack."))
+			stack.stackmob.verbs |= /mob/proc/resleeve
+			stack.stackmob.verbs |= /mob/proc/death_give_up
+		else
+			to_world_log("no stack found")
```
Comment (Andrew-Fall) [inline mods/persistence/modules/mob/death.dm]: Debug message

File: mods/persistence/modules/mob/death.dm [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+			// Check if they have *any* backups.
+			if(length(get_valid_clone_pods(mind.unique_id))) // we have a backup.
+				to_world_log("putting in limbo.")
```
Comment (Andrew-Fall) [inline mods/persistence/modules/mob/death.dm]: Debug message

File: mods/persistence/modules/mob/death.dm [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+				brainmob.SetName(real_name)
+				brainmob.real_name = real_name
+				brainmob.dna = dna.Clone()
+				brainmob.verbs |= /mob/proc/resleeve
+				brainmob.verbs |= /mob/proc/death_give_up
+				// brainmob.timeofhostdeath = timeofdeath
+				mind.transfer_to(brainmob)
+				to_chat(brainmob, SPAN_NOTICE("Darkness envelopes you. Your character has died and you are now in limbo. Your sleeve has been destroyed but you still have backups in the world. Resleeve to continue playing as your character."))
+			else
+				to_world_log("new player time")
```
Comment (Andrew-Fall) [inline mods/persistence/modules/mob/death.dm]: Debug message

File: mods/persistence/modules/organs/internal/stack.dm [hunk 11]
```diff
@@ -1,29 +1,11 @@
...
-	if(key)
-		crash_with("TODO: Switch to backup or destroy.")
-	. = ..()
+/proc/switchToStack(var/datum/mind/mind)
+	for(var/obj/item/organ/internal/stack/S in world)
+		if(S.owner.mind == mind)
+			S.stackmob = new()
+			S.stackmob.forceMove(S)
+			mind.transfer_to(S.stackmob)
+			to_chat(S.stackmob, "<span class='notice'>You feel slightly disoriented. That's normal when you're just \a [S.name].</span>")
```
Comment (Andrew-Fall) [inline mods/persistence/modules/organs/internal/stack.dm]: SPAN_NOTICE

File: mods/persistence/modules/organs/internal/stack.dm [hunk 12]
```diff
@@ -61,75 +37,27 @@ GLOBAL_LIST_EMPTY(corticalStacks)
 		to_chat(user, "The light on [src] is off. " + (user.skill_check(SKILL_DEVICES, SKILL_EXPERT) ? "It doesn't have a backup." : "Wonder what that means."))
 
 /obj/item/organ/internal/stack/emp_act()
-	return
+	return FALSE
 
 /obj/item/organ/internal/stack/getToxLoss()
-	return 0
+	return FALSE
```
Comment (Andrew-Fall) [inline mods/persistence/modules/organs/internal/stack.dm]: Seems like it should remain 0

**Review by Andrew-Fall — CHANGES_REQUESTED** [review]

File: mods/persistence/modules/cloning/backup_file.dm (hunk 4 above)
Comment (biscuitWizard) [reply mods/persistence/modules/cloning/backup_file.dm:6]: Separate PR.

File: mods/persistence/modules/cloning/machines/cloning_pod.dm [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+	set src in oview(1)
+
+	attempt_enter(usr, usr, "\The [usr] starts climbing into \the [src].")
+
+/obj/machinery/cloning_pod/proc/attempt_enter(var/atom/movable/target, var/mob/user, var/message)
+	if(user.stat != 0 || !check_occupant_allowed(target))
+		return
+
+	if(occupant)
+		to_chat(user, "<span class='notice'><B>\The [src] is in use.</B></span>")
```
Comment (Andrew-Fall) [inline mods/persistence/modules/cloning/machines/cloning_pod.dm]: SPAN_NOTICE

File: mods/persistence/modules/organs/internal/stack.dm (hunk 11 above)
Comment (Andrew-Fall) [inline mods/persistence/modules/organs/internal/stack.dm]: Still SPAN_NOTICE

**Review by Andrew-Fall — CHANGES_REQUESTED** [review]

**Review by Andrew-Fall — APPROVED** [review]

**Pull request merged**